src/core/fetchers/segment/promise_prioritizer.ts

```diff
@@ -0,0 +1,311 @@
+import log from "../../../log";
+import arrayFindIndex from "../../../utils/array_find_index";
+import PPromise from "../../../utils/promise";
+import {
+  CancellationError,
+  CancellationSignal,
+} from "../../../utils/task_canceller";
+
+export default class PromisePrioritizer<T> {
+  private _waitingQueue: Array<IPrioritizerTask<T>>;
+  private _pendingTasks: Array<IPrioritizerTask<T>>;
+  private _prioritySteps: IPrioritizerPrioritySteps;
+
+  constructor({ prioritySteps }: IPrioritizerOptions) {
+    this._waitingQueue = [];
+    this._pendingTasks = [];
+    this._prioritySteps = prioritySteps;
+
+    if (this._prioritySteps.high >= this._prioritySteps.low) {
+      throw new Error(
+        "FP Error: the max high level priority should be given a lower" +
+          "priority number than the min low priority."
+      );
+    }
+  }
+
+  private get _minPendingPriority(): number | null {
```

Comment:
@peaBerberian I changed the minPendingPriority to a getter to simplify handling it in the code

Reply:
Why does it simplify the code?
I'm not too much of a fan of getters in those kind of cases because there's a lot of things happening here (Math.min, `map` on pendingTasks ...) where a caller would assume that only a value would be read.

Reply:
There were a lot of occurence were we do complicated computation and checks to end up exactly the same as doing ```minpendingpriorrity = pendingTask.length === 0  ? null : Math.min(pendingTask.map(t => t.priority))``` , we were doing 2/3 checks each time , in almost every operation( is the task that we are dealing with no longer in pendingTask, is the pendingTasks empty, is the task we just deleted the highest priority there was ) that were not useful to me , as just taking null if there is no task , else the min priority sums up all those cases and checks.
So that code just got remplaced with this 3 lines.